packages/run-protocol/src/vaultFactory/vaultManager.js

```diff
@@ -151,48 +204,122 @@ export const makeVaultManager = (
-        liquidationStrategy,
-        collateralBrand,
-      ),
+  // In extreme situations, system health may require liquidating all vaults.
+  const liquidateAll = async () => {
+    assert(prioritizedVaults);
+    const toLiquidate = Array.from(prioritizedVaults.entries()).map(
+      liquidateAndRemove,
     );
-    return Promise.all(promises);
+    await Promise.all(toLiquidate);
   };
 
+  /**
+   *
+   * @param {bigint} updateTime
+   * @param {ZCFSeat} poolIncrementSeat
+   */
   const chargeAllVaults = async (updateTime, poolIncrementSeat) => {
-    const poolIncrement = sortedVaultKits.reduce(
-      (total, vaultPair) =>
-        AmountMath.add(
-          total,
-          vaultPair.vaultKit.accrueInterestAndAddToPool(updateTime),
-        ),
-      AmountMath.makeEmpty(runBrand),
+    trace('chargeAllVault', { updateTime });
+    const interestCalculator = makeInterestCalculator(
+      shared.getInterestRate(),
+      shared.getChargingPeriod(),
+      shared.getRecordingPeriod(),
+    );
+
+    // calculate delta of accrued debt
+    const debtStatus = interestCalculator.calculateReportingPeriod(
+      {
+        latestInterestUpdate,
+        newDebt: totalDebt,
+        interest: 0n, // XXX this is always zero, doesn't need to be an option
+      },
+      updateTime,
+    );
+    const interestAccrued = debtStatus.interest;
+
+    // done if none
+    if (interestAccrued === 0n) {
+      return;
+    }
+
+    // NB: This method of inferring the compounded rate from the ratio of debts
+    // acrrued suffers slightly from the integer nature of debts. However in
+    // testing with small numbers there's 5 digits of precision, and with large
+    // numbers the ratios tend towards ample precision. Because this calculation
+    // is over all debts of the vault the numbers will be reliably large.
+    compoundedInterest = calculateCompoundedInterest(
+      compoundedInterest,
+      totalDebt,
+      debtStatus.newDebt,
     );
-    sortedVaultKits.updateAllDebts();
+    totalDebt += interestAccrued;
+
+    // mint that much RUN for the reward pool
+    const rewarded = AmountMath.make(runBrand, interestAccrued);
+    runMint.mintGains(harden({ RUN: rewarded }), poolIncrementSeat);
+    reallocateReward(rewarded, poolIncrementSeat);
+
+    // update running tally of total debt against this collateral
+    ({ latestInterestUpdate } = debtStatus);
+
+    const payload = harden({
+      compoundedInterest,
+      latestInterestUpdate,
+      totalDebt: AmountMath.make(runBrand, totalDebt),
+    });
+    updater.updateState(payload);
+
+    trace('chargeAllVaults complete', payload);
+
     reschedulePriceCheck();
-    runMint.mintGains(harden({ RUN: poolIncrement }), poolIncrementSeat);
-    reallocateReward(poolIncrement, poolIncrementSeat);
+  };
+
+  /**
+   * @param {Amount<NatValue>} oldDebtOnVault
+   * @param {Amount<NatValue>} newDebtOnVault
+   */
+  const applyDebtDelta = (oldDebtOnVault, newDebtOnVault) => {
```

Comment:
Please add a ticket to consider limits on this. At the very least, it should never be more than the max allowed for a vault, for example.

Reply:
And it should never be larger than this manager's total debt (so a bug in one manager can screw up the debt limit on the overall service.

Reply:
https://github.com/Agoric/agoric-sdk/issues/4599